src/konserve/indexeddb.cljs

```diff
@@ -1,281 +1,316 @@
-            [konserve.serializers :as ser]))
+  (:require [cljs.core.async :refer [go take! put! close!]]
+            [konserve.compressor]
+            [konserve.encryptor]
+            [konserve.impl.defaults :as defaults]
+            [konserve.impl.storage-layout :as storage-layout :refer [PBackingStore]]
+            [konserve.serializers]
+            [konserve.utils :refer [with-promise]]))
 
-#_(defrecord IndexedDBKeyValueStore [db store-name serializer read-handlers write-handlers locks version]
-    PEDNKeyValueStore
-    (-exists? [_this key]
-      (let [res       (chan)
-            tx        (.transaction db #js [store-name])
-            obj-store (.objectStore tx store-name)
-            req       (.openCursor obj-store (pr-str key))]
-        (set! (.-onerror req)
-              (fn [e]
-                (put! res (ex-info "Cannot check for existence."
-                                   {:type  :access-error
-                                    :key   key
-                                    :error (.-target e)}))
-                (close! res)))
-        (set! (.-onsuccess req)
-              (fn [e]
-                (put! res (if (.-result (.-target e))
-                            true false))
-                (close! res)))
-        res))
-
-    (-get-meta [_this key]
-      (let [res       (chan)
-            tx        (.transaction db #js [store-name])
-            obj-store (.objectStore tx store-name)
-            req       (.get obj-store (pr-str key))]
-        (set! (.-onerror req)
-              (fn [e]
-                (put! res (ex-info "Cannot read edn value."
-                                   {:type  :read-error
-                                    :key   key
-                                    :error (.-target e)}))
-                (close! res)))
-        (set! (.-onsuccess req)
-              (fn [e] (when-let [r (.-result req)]
-                        (put! res (-deserialize serializer read-handlers (aget r "meta"))))
-                (close! res)))
-        res))
-
-    (-get [this key]
-      (let [res       (chan)
-            tx        (.transaction db #js [store-name])
-            obj-store (.objectStore tx store-name)
-            req       (.get obj-store (pr-str key))]
-        (set! (.-onerror req)
-              (fn [e]
-                (put! res (ex-info "Cannot read edn value."
-                                   {:type  :read-error
-                                    :key   key
-                                    :error (.-target e)}))
-                (close! res)))
-        (set! (.-onsuccess req)
-              (fn [e] (when-let [r (.-result req)]
-                        (put! res (-deserialize serializer read-handlers (aget r "edn_value"))))
-                (close! res)))
-        res))
+(defn- connect-to-idb [db-name]
+  (let [req (js/window.indexedDB.open db-name 1)]
+    (with-promise out
+      (set! (.-onblocked req) (fn [event] (put! out [event])))
```

Comment:
The way we handle errors in https://github.com/replikativ/superv.async/ that we use on top of this API is not to put a return value and error into a tuple, but instead disambiguate them in `<?` by type and rethrow exceptions. This might have downsides (e.g. performance), but it is a nice compositional error handling that does not require manual creations of `[out err]` everywhere. For this reason the expected return type should not be tupled. I am happy to discuss improving `superv.async` though.

Reply:
FWIW I took care to make sure all public apis follow superv async conventions but generally I have learned to use go blocks sparingly because they are compilation land mines whereas destructing argvecs are always free. I am happy to change it if you lik

Reply:
I understand. Connecting to a store should behave the same everywhere though, so it would be better here to just return either the error or the value on the same promise/channel.